diff --git a/src/main/java/net/minecraft/server/WorldNBTStorage.java b/src/main/java/net/minecraft/server/WorldNBTStorage.java
index 551376ff..eaae4468 100644
--- a/src/main/java/net/minecraft/server/WorldNBTStorage.java
+++ b/src/main/java/net/minecraft/server/WorldNBTStorage.java
@@ -151,24 +151,21 @@ public class WorldNBTStorage implements IPlayerFileData {
     }
 
     public void checkSession() throws ExceptionWorldConflict {
-        // MIMC: Disable this check to enable multiple instances dealing with the same world concurrently
-        return;
-
-//        try {
-//            File file = new File(this.baseDir, "session.lock");
-//            DataInputStream datainputstream = new DataInputStream(new FileInputStream(file));
-//
-//            try {
-//                if (datainputstream.readLong() != this.sessionId) {
-//                    throw new ExceptionWorldConflict("The save for world located at " + this.baseDir + " is being accessed from another location, aborting");  // Spigot
-//                }
-//            } finally {
-//                datainputstream.close();
-//            }
-//
-//        } catch (IOException ioexception) {
-//            throw new ExceptionWorldConflict("Failed to check session lock for world located at " + this.baseDir + ", aborting. Stop the server and delete the session.lock in this world to prevent further issues."); // Spigot
-//        }
+        try {
+            File file = new File(this.baseDir, "session.lock");
+            DataInputStream datainputstream = new DataInputStream(new FileInputStream(file));
+
+            try {
+                if (datainputstream.readLong() != this.sessionId) {
+                    throw new ExceptionWorldConflict("The save for world located at " + this.baseDir + " is being accessed from another location, aborting");  // Spigot
+                }
+            } finally {
+                datainputstream.close();
+            }
+
+        } catch (IOException ioexception) {
+            throw new ExceptionWorldConflict("Failed to check session lock for world located at " + this.baseDir + ", aborting. Stop the server and delete the session.lock in this world to prevent further issues."); // Spigot
+        }
     }
 
     @Nullable
