commit ea6b56ad30c3681b0cc102c8c9df3852b2e4a81a
Author: Yannick Motton <ymotton@gmail.com>
Date:   Thu Aug 8 03:25:46 2019 +0200

    Seamless transitions

diff --git a/proxy/src/main/java/net/md_5/bungee/connection/DownstreamBridge.java b/proxy/src/main/java/net/md_5/bungee/connection/DownstreamBridge.java
index 845d6a33..f8d9f760 100644
--- a/proxy/src/main/java/net/md_5/bungee/connection/DownstreamBridge.java
+++ b/proxy/src/main/java/net/md_5/bungee/connection/DownstreamBridge.java
@@ -138,23 +138,23 @@ public class DownstreamBridge extends PacketHandler
         return !server.isObsolete();
     }
 
-    private void updatePlayerLocation(PacketWrapper packet) {
+    private boolean updatePlayerLocation(PacketWrapper packet) {
         double x = packet.buf.readDouble();
         double y = packet.buf.readDouble();
         double z = packet.buf.readDouble();
         float yaw = packet.buf.readFloat();
         float pitch = packet.buf.readFloat();
         byte flags = packet.buf.readByte();
-        //bungee.getLogger().info( String.format("PacketPlayOutPlayerPositionAndLook %s X: %.2f Y: %.2f Z: %.2f", con.getUniqueId(), x, y, z) );
+        //bungee.getLogger().info( String.format("PacketPlayOutPlayerPositionAndLook %s X: %.2f Y: %.2f Z: %.2f", con.getName(), x, y, z) );
 
-        ((BungeeCord)bungee).partitionManager.movePlayer(
+        return ((BungeeCord)bungee).partitionManager.tryTeleportPlayer(
                 con.getUniqueId()
                 , x, (flags & 0x1) == 0x1
                 , y, (flags & 0x2) == 0x2
                 , z, (flags & 0x4) == 0x4
                 , yaw
                 , pitch
-            );
+        );
     }
 
     @Override
@@ -223,7 +223,7 @@ public class DownstreamBridge extends PacketHandler
 //            case 0x34: packetName = "PacketPlayOutFacePlayer"; break;
             case 0x35:
 //                packetName = "PacketPlayOutPlayerPositionAndLook";
-                updatePlayerLocation(packet);
+                if (!updatePlayerLocation(packet)) return;
                 break;
 //            case 0x36: packetName = "PacketPlayOutUnlockRecipes"; break;
 //            case 0x37: packetName = "PacketPlayOutEntityDestroy"; break;
diff --git a/proxy/src/main/java/net/md_5/bungee/mimc/MigrationRequest.java b/proxy/src/main/java/net/md_5/bungee/mimc/MigrationRequest.java
index 42f36490..e8d5324b 100644
--- a/proxy/src/main/java/net/md_5/bungee/mimc/MigrationRequest.java
+++ b/proxy/src/main/java/net/md_5/bungee/mimc/MigrationRequest.java
@@ -3,10 +3,18 @@ package net.md_5.bungee.mimc;
 public class MigrationRequest {
     public final Player player;
     public final Partition target;
+    public final Long requestedAt;
+    public Long completedAt;
 
     public MigrationRequest(Player player, Partition target) {
         this.player = player;
         this.target = target;
+        this.requestedAt = System.currentTimeMillis();
+    }
+
+    public MigrationRequest complete() {
+        completedAt = System.currentTimeMillis();
+        return this;
     }
 
     public int hashCode() {
diff --git a/proxy/src/main/java/net/md_5/bungee/mimc/PartitionManager.java b/proxy/src/main/java/net/md_5/bungee/mimc/PartitionManager.java
index f23052cb..48cfc636 100644
--- a/proxy/src/main/java/net/md_5/bungee/mimc/PartitionManager.java
+++ b/proxy/src/main/java/net/md_5/bungee/mimc/PartitionManager.java
@@ -27,6 +27,7 @@ public class PartitionManager implements Listener {
     private List<Partition> partitions = new ArrayList<>();
     private List<Player> players = new ArrayList<>();
     private CopyOnWriteSet<MigrationRequest> migrationRequests = new CopyOnWriteSet<>();
+    private CopyOnWriteSet<MigrationRequest> completedMigrations = new CopyOnWriteSet<>();
 
     public PartitionManager(BungeeCord bungeeCord) {
         this.bungeeCord = bungeeCord;
@@ -88,13 +89,23 @@ public class PartitionManager implements Listener {
         }
         return player;
     }
-    public void movePlayer(UUID uuid
+    public boolean tryTeleportPlayer(UUID uuid
             , double x, boolean xIsRelative
             , double y, boolean yIsRelative
             , double z, boolean zIsRelative
             , float yaw, float pitch) {
         Player player = getPlayer(uuid);
         if (player != null) {
+            MigrationRequest request = getLatestMigration(player.name);
+            if (request != null) {
+                // Grace period after a new player connecting that we do not proxy teleports
+                // This eliminates the rubber banding issue when migrating between nodes
+                if (System.currentTimeMillis() < request.completedAt + 5_000) {
+                    bungeeCord.getLogger().warning(String.format("MIMC | Player %s Teleport ignored during grace period after migration.", player.name));
+                    return false;
+                }
+            }
+
             player.move(
                     xIsRelative ? player.location.calculateOffsetFromX(x) : x
                     , yIsRelative ? player.location.calculateOffsetFromY(y) : y
@@ -104,6 +115,7 @@ public class PartitionManager implements Listener {
 
             migratePlayer(player);
         }
+        return true;
     }
 
     public Player addPlayer(UserConnection userCon) {
@@ -157,44 +169,63 @@ public class PartitionManager implements Listener {
         sendMigrationMessage(p, target);
     }
 
-    public boolean completeMigration(String playerName) {
-        bungeeCord.getLogger().warning( String.format("MIMC | completeMigration for %s", playerName) );
-
+    MigrationRequest getLatestMigrationRequest(String playerName) {
         Iterator<MigrationRequest> iterator = this.migrationRequests.iterator();
+        MigrationRequest latestRequest = null;
+        while(iterator.hasNext()) {
+            MigrationRequest request = iterator.next();
+            if (request.player.name.equals(playerName)) {
+                if (latestRequest == null) {
+                    latestRequest = request;
+                } else {
+                    if (latestRequest.requestedAt < request.requestedAt) {
+                        latestRequest = request;
+                    }
+                }
+            }
+        }
+        return latestRequest;
+    }
+    MigrationRequest getLatestMigration(String playerName) {
+        Iterator<MigrationRequest> iterator = this.completedMigrations.iterator();
+        MigrationRequest latestRequest = null;
         while(iterator.hasNext()) {
             MigrationRequest request = iterator.next();
+            if (request.player.name.equals(playerName)) {
+                if (latestRequest == null) {
+                    latestRequest = request;
+                } else {
+                    if (latestRequest.completedAt < request.completedAt) {
+                        latestRequest = request;
+                    }
+                }
+            }
+        }
+        return latestRequest;
+    }
+
+    public boolean completeMigration(String playerName) {
+        bungeeCord.getLogger().warning( String.format("MIMC | completeMigration for %s", playerName) );
+
+        MigrationRequest request = getLatestMigrationRequest(playerName);
+        if (request != null) {
             Player p = request.player;
             Partition target = request.target;
 
-            if (p.name.equals(playerName)) {
-                p.migrateTo(target);
-
-                p.userCon.connect(servers.get(target.serverName), (result, error) -> {
-                    if (result) {
-                        // Immediately close loading screen
-//                        ByteArrayDataOutput out = ByteStreams.newDataOutput();
-//                        out.writeByte(0x35);
-//                        out.writeDouble(p.location.x);
-//                        out.writeDouble(p.location.y);
-//                        out.writeDouble(p.location.z);
-//                        out.writeFloat(p.location.yaw);
-//                        out.writeFloat(p.location.pitch);
-//                        out.writeByte(0); // Flags
-//                        out.writeByte(0); // Teleport ID
-//                        ByteBuf buf = Unpooled.wrappedBuffer(out.toByteArray());
-//                        PacketWrapper packet = new PacketWrapper( null, buf );
-//                        p.userCon.sendPacket(packet);
-
-                    } else {
-                        String errorMessage = "Failed to connect " + p.userCon.getName() + " to " + request.target.getServerName();
-                        this.bungeeCord.getLogger().log(Level.WARNING, errorMessage, error);
-                        p.userCon.disconnect(errorMessage);
-                    }
-                });
+            p.migrateTo(target);
 
-                this.migrationRequests.remove(request);
-                return true;
-            }
+            p.userCon.connect(servers.get(target.serverName), (result, error) -> {
+                if (result) {
+                    completedMigrations.add(request.complete());
+                } else {
+                    String errorMessage = "Failed to connect " + p.userCon.getName() + " to " + request.target.getServerName();
+                    this.bungeeCord.getLogger().log(Level.WARNING, errorMessage, error);
+                    p.userCon.disconnect(errorMessage);
+                }
+            });
+
+            this.migrationRequests.remove(request);
+            return true;
         }
         return false;
     }
@@ -222,6 +253,7 @@ public class PartitionManager implements Listener {
 
         // Gets the server with the highest TPS
         String serverWithHighestTPS = this.serverTPSMap.entrySet().stream()
+                .filter(x -> !x.getKey().equals(SPAWN_SERVERNAME))
                 .max(Comparator.comparing(Map.Entry::getValue))
                 .get().getKey();
 
